hack/run-functional-tests.sh

```diff
@@ -73,107 +73,171 @@ check_arguments() {
 
-                  elif [ "${test_type}"  == "fail" ] && [ ${exit_code} -ne 1 ]; then
+                else
 
-                      echo "[ERROR] Test did not pass. Exiting..."
-                      exit 1
+                    echo "[INFO] Test Passed"
 
-                  else
+                fi
 
-                      echo "[INFO] Test Passed"
+            fi
 
-                  fi
+        else
 
-              fi
+            echo "[WARN] File \"${test_file_path}\" not found. Skipping..."
 
-          else
+        fi
 
-              echo "[WARN] File \"${test_file_path}\" not found. Skipping..."
+        # check to see if the user specified a script to be associated with this stanza
+        # only run script for apply actions
+        if [[ ${user_script} != "" ]] && [[ "${action}" == "apply" ]]; then
 
-          fi
+          # if they did specify a script run it with the between argument and pass the namespace in
+          ${user_script_path} "between" "${TEST_NAMESPACE}"
 
-          echo "============================================================================"
-          
-      done
+        fi
+
+        echo "============================================================================"
+        
+    done
+
+    # check to see if the user specified a script to be associated with this stanza
+    # only run script for apply actions
+    if [[ ${user_script} != "" ]] && [[ "${action}" == "apply" ]]; then
+
+      # if they did specify a script run it with the teardown argument and pass the namespace in
+      ${user_script_path} "teardown" "${TEST_NAMESPACE}"
+
+      echo "============================================================================"
 
     fi
 
+  fi
+
 }
 
 # **********************************************
 # Determine test scope
 # **********************************************
 scope_and_run_tests() {
 
+  #create identifiable local variable for argument 1, the action to perform
   local action="${1}"
 
-  if [ "${TEST_RESOURCE_TYPE}" == "all" ]; then
-
-    resources=$(yq read "${TESTS_MANIFEST}" 'resources.[*].name')
+  #size the array of resources
+  local resource_array_length
+  resource_array_length=$(yq read -l "${TESTS_MANIFEST}" 'resources')
 
-    for resource in ${resources}; do
 
-      if [ "${TEST_TYPE}" == "all" ]; then
+  #loop through all resources in the supplied manifest
+    #determine which indicies meet the supplied criteria in ${TEST_RESOURCE_KIND} and ${TEST_RESOURCE_DESIRED}
+  for ((i = 0 ; i < resource_array_length ; i++)); do
 
-        run_resource_tests "${action}" "${resource}" "pass"
-        run_resource_tests "${action}" "${resource}" "fail"
+    #check if we're doing all resources or if the resource kind at $i matches the requested kind
+    #double brackets are technically correct; the BEST kind of correct!
+    if [[ "${TEST_RESOURCE_KIND}" == "all" ]] || [[ "${TEST_RESOURCE_KIND}" == $(yq read "${TESTS_MANIFEST}" "resources.[${i}].kind") ]]; then
```

Comment:
```suggestion
    if [[ "${TEST_RESOURCE_KIND}" == "all" ]] || [[ "${TEST_RESOURCE_KIND}" == "$(yq read "${TESTS_MANIFEST}" "resources.[${i}].kind")" ]]; then
```

Reply:
I'm not sure on this one to be honest. I did not add the extra double quotes on this because I think it makes a weird result of what is and isn't quoted. "$(yq read " is quoted but then  ${TESTS_MANIFEST} is unquoted and so forth because they just pair up they don't nest inside each other right?

Reply:
I believe the parenthesis in this case control the expansion in order, so the double quotes around the parens are evaluated after the `yq` command is expanded. This is called indirect expansion if I recall.

Example:

```shell
$ test1="test1"
$ echo "$(echo "${test1}-123")"
$ test1-123
```

Reply:
OK, we'll go with that then.